### PR TITLE
AGENT-227: Add host configuration support to ephemeral installer client

### PR DESCRIPTION
**Pull request by zaneb**
* [AGENT-238](https://issues.redhat.com/browse/AGENT-238) Add subcommands to the ephemeral installer client
* [AGENT-227](https://issues.redhat.com/browse/AGENT-227) Add support for configuring the install disk based on root device hints
* [AGENT-229](https://issues.redhat.com/browse/AGENT-229) Add support for configuring the role for each host

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] Ephemeral installer
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @flaper87
/cc @lranjbar

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by mkowalski** [comment]
I reviewed the part covering configuring the install disk based on root device hints and it looks okay for me. Unfortunately at this moment I don't have enough knowledge about the rest of the code touched by this PR to safely LGTM it as a whole

**Comment by celebdor** [comment]
This PR should get /lgtm and /approval from approvers in ZTP side (due to the rootdevicehints). Let's put a hold until we have /lgtm from agent side and /lgtm from ZTP side and /approve from somebody on the approvers list.

File: internal/controller/controllers/bmh_agent_controller.go
Comment (paul-maidment) [inline internal/controller/controllers/bmh_agent_controller.go:916]: Seems reasonable to pick the first acceptable disk though I do question whether or not there might need to be consideration of which of the acceptable is "best"

Are there any other priorities under which we would like to select the acceptable disk, is the first always the best choice?
Should the list of acceptable disks be sorted any way to place the "most acceptable" first?

File: internal/host/hostutil/host_utils.go
Comment (paul-maidment) [inline internal/host/hostutil/host_utils.go:171]: As it stands in the current implementation, hints will never be nil due to the check in `findInstallationDiskID` is the desired behaviour that findInstallationDiskID in the absence of hints should return an arbitrary disk ID or should it return an empty ID.

If the former, then should the check be removed from `findInstallationDiskID` and if the latter should this check be removed here?

Comment (paul-maidment) [reply internal/host/hostutil/host_utils.go:171]: Actually, noted that a few lines below there is an eligibility check for the disk, which will only run if this condition is not triggered. As it stands, this function could return ineligible disks in the collection if there are no hints specified.

Comment (zaneb) [reply internal/host/hostutil/host_utils.go:171]: > Is the desired behaviour that findInstallationDiskID in the absence of hints should return an arbitrary disk ID or should it return an empty ID?

For the existing BMAC use, it should return nothing. For the new use in this PR, it's easier if it returns the list of all disks. Hence the checks in both places.

> Actually, noted that a few lines below there is an eligibility check for the disk, which will only run if this condition is not triggered. As it stands, this function could return ineligible disks in the collection if there are no hints specified.

That's a good point, thanks.

**Review by paul-maidment — APPROVED** [review]
/lgtm

**Comment by flaper87** [comment]
/lgtm
/approve

The rootdevicehints piece looks sane!

File: internal/controller/controllers/bmh_agent_controller.go
Comment (zaneb) [reply internal/controller/controllers/bmh_agent_controller.go:898]: I had that originally, but it seemed wrong to cast a string that was possibly in the wrong case to a `DriveType`. Didn't notice the multipath one though tbh. In practice we're only checking for `"HDD"` so it will work. Not sure what the least wrong thing is.

**Pull request merged**
